[FLINK-17398][connector/filesystem] Filesystem sources support flexible path reading

ruanhang1993 opened the pull request:
## What is the purpose of the change

This pull request makes filesystem sources support flexible path reading.

## Brief change log

  - Add two new FileEnumerator and a new file filter.
  - Add new table option 'source.regex-pattern'
  - use the new FileEnumerator if 'source.regex-pattern' is set

## Verifying this change

This change added tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not documented

JuliaBogdan commented:
Hi @ruanhang1993, these changes are very much needed, thank you!
Are there any blockers for the PR to be reviewed and merged?

Also, is there a workaround to this for the meantime, i.e. a way to specify a custom enumerator to `FileSource` or a custom file filter to the existing enumerators?

rculbertson commented:
I'd also like to see this PR merged, so gave this a +1 as vote of support :)

RamanVerma commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/FileSystemTableSource.java:
```diff
@@ -269,10 +271,35 @@ private SourceProvider createSourceProvider(BulkFormat<RowData, FileSourceSplit>
+            result.append(Path.SEPARATOR);
+        }
+        int startIndex = 0;
+        while (startIndex < regex.length()
```
I think this loop is removing multiple path separators at the start of regex, for example `////.*`
Can there be a regex like this `////.*///.*`

luoyuxia replied:
Still, I'm try to understand why need `connectBasePathAndRegex`. It turns out that what the option `source.regex-pattern` is for?
 From the source code, seems it'll look like if the table path is `/dir`,  the `source.regex-pattern` is `t/*`, it will try to match `/dir/t/*`?

TBH, I'm not sure it's reasonable or not for it look wired or unclear to me.
Another choice is user need to specifc `/dir/t/*` in `source.regex-pattern`.

I don't have a strong perference about these two choice, and   don't know which is a better choice. Let's hear other's voice.

ruanhang1993 replied:
@luoyuxia Using a base path will help to reduce the search time. 
If we only provide `source.regex-pattern` , we have to search files from root path `/` when `source.regex-pattern` = `'/dir/t/.*/inner/.*'`.
If we provide base dir and `source.regex-pattern`, we will start to search files under the base dir `/dir/t/` when base dir = `'/dir/t'` and `source.regex-pattern` = `'.*/inner/.*'`.

luoyuxia replied:
No, I mean user provide base dir and `source.regex-pattern`. But the `source.regex-pattern` should also be absolute path insteading of relative path.  But we only serach from the base dir  and to check each file.

Take this for example, `source.regex-pattern` will be `/dir/t/.*/inner/.*` which looks more clear than `.*/inner/.*` to me.
I'm wondering whether use `/dir/t/.*/inner/.*` will be more clear.

RamanVerma reviewed (commented):
thanks for the PR, @ruanhang1993 . I have left some minor comments

ruanhang1993 commented:
Hi, @RamanVerma . Thanks for your review. I will rebase the master and fix comments later.

luoyuxia commented:
Thanks all. I'll reivew & merge. 
I hope others can also have a look to make sure this pr can really meet your need; :)

ruanhang1993 commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/enumerate/NonSplittingRegexEnumerator.java:
```diff
@@ -0,0 +1,90 @@
+        for (Path path : paths) {
+            final FileSystem fs = path.getFileSystem();
+            final FileStatus status = fs.getFileStatus(path);
+            addSplitsForPath(status, fs, splits, regexFileFilter);
```
The method `addSplitsForPath ` will be invoked in itself at line 83. The last parameter may be null.
When it is null, we will know the dir/file comes from a matched path. And we will handle them if they are not filtered by the custom `fileFilter`.

ruanhang1993 replied:
It aims to matching a directory and handling all files under the matched directory.

ruanhang1993 commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/enumerate/NonSplittingRegexEnumerator.java:
```diff
@@ -0,0 +1,90 @@
+ */
+public class NonSplittingRegexEnumerator extends NonSplittingRecursiveEnumerator {
+
+    private final Predicate<Path> fileFilter;
```
I will add some comments about them.

JuliaBogdan replied:
A bit out of scope for the PR, but wanted to get people's thoughts on this:
> allows caller to pass in custom filters

If I'm not mistaken, there is no way to specify a custom file filter for `FileSystemTableSource`  in the current implementation (neither via custom file enumerator or custom predicate to the existing ones).
Is it by design? If so, what would be an option for passing it in, except implementing a custom version of `FileSystemTableSource`?

ruanhang1993 replied:
I think this custom filter may not be exposed to the table API. We can only use it by the DataStream API.

ruanhang1993 commented on flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/src/enumerate/NonSplittingRegexEnumeratorTest.java:
```diff
@@ -0,0 +1,247 @@
+    }
+
+    @Test
+    void testFileWithIncorrectBlocks() throws Exception {
```
This test also lies in the `NonSplittingRecursiveEnumeratorTest`.
I think `NonSplittingRecursiveEnumerator` and `NonSplittingRegexEnumerator` could also deal with the files with blocks.

luoyuxia commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/enumerate/NonSplittingRegexEnumerator.java:
```diff
@@ -0,0 +1,102 @@
+    /** The custom filter predicate to filter out unwanted files. */
+    private final Predicate<Path> fileFilter;
+
+    private final RegexFileFilter regexFileFilter;
```
Do we really need this in here? I think in here `fileFilter` is enough. We can always construnct the `fileFilter` using regex expression.

luoyuxia replied:
I mean construct the `fileFilter`  in `FileSystemTableSource`.

luoyuxia replied:
Also, I'm now thing do we really need to introduce `NonSplittingRecursiveEnumerator`?
Why don't reuse `NonSplittingRecursiveEnumerator`, but the `fileFilter` will compose the logic of `DefaultFileFilter` and `RegexFileFilter`.

ruanhang1993 replied:
Because it does not satisfy some situations. 
For example, we have files as follows.
```
/
 |- dir
     |- test
           |- test.java
           |- innerDir
                 |- test.java
```
We want to read these two `test.java` under the directory `/dir`. Then we will use the final regex `/dir/.*/test.java`.
If we reuse `NonSplittingRecursiveEnumerator ` + `fileFilter`, it will skip the directory `/dir/test` because its path does not match the regex `/dir/.*/test.java`. And it will lose the files.

ruanhang1993 replied:
I think it is not suitable when the regex is complex. I think I should remove the `connectBasePathAndRegex ` and use the actual value passed by `source.regex-pattern`.

But I still think we should keep the table path as the base dir to start to search files.

luoyuxia reviewed (commented):
@ruanhang1993 Thanks for contribution. I left some comments.

ruanhang1993 commented:
Hi, @luoyuxia & @RamanVerma .
I have refactored this PR and removed the `connectBasePathAndRegex`. The previous version is hard to understand and use . This new code will use the actual regex string to match all files in the base dir.

Please take a look at it . Thanks ~

luoyuxia commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/connector/file/table/FileSystemTableSourceStreamingITCase.java:
```diff
@@ -88,4 +88,61 @@ public void testMonitorContinuously() throws Exception {
     }
+
+    @Test
+    public void testRegexPath() throws Exception {
```
I think the test can't verify the `source.regex-pattern` take effect or not since it won't filer out any files.

luoyuxia replied:
nit
```suggestion
    public void testSourceWithRegexPattern() throws Exception {
```

luoyuxia reviewed (commented):
@ruanhang1993 Thanks for updating. The pr is in good shape.
I left some comments. PTAL.

luoyuxia commented:
Also, don't forget to update doc for the newly added options. I'm fine with adding it in this pr or another pr.

ruanhang1993 commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/enumerate/BlockSplittingRecursiveAllDirEnumerator.java:
```diff
@@ -0,0 +1,165 @@
+ * <p>Compared to {@link BlockSplittingRecursiveEnumerator}, this enumerator will enumerate all
+ * files even through its parent directory is filtered out by the file filter.
+ */
+public class BlockSplittingRecursiveAllDirEnumerator extends NonSplittingRecursiveAllDirEnumerator {
```
`BlockSplittingRecursiveAllDirEnumerator` need to reuse the method `addSplitsForPath` in `NonSplittingRecursiveAllDirEnumerator`, not `BlockSplittingRecursiveEnumerator`.

luoyuxia replied:
If reuse `addSplitsForPath`  in `NonSplittingRecursiveAllDirEnumerator`, we can't resue the code in `BlockSplittingRecursiveEnumerator`.
So, it comes out whether to resue the code in `BlockSplittingRecursiveEnumerator` or the code in `NonSplittingRecursiveAllDirEnumerator `.  
Why not choose to reuse the code in `BlockSplittingRecursiveEnumerator`? Otherwise, we need to rewrite the codes in  `BlockSplittingRecursiveEnumerator` which is a little of complex to me.

ruanhang1993 commented:
> Also, don't forget to update doc for the newly added options. I'm fine with adding it in this pr or another pr.

Thanks for reviewing. I will add docs in this PR.

luoyuxia commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/enumerate/BlockSplittingRecursiveAllDirEnumerator.java:
```diff
@@ -0,0 +1,97 @@
+            }
+            final FileStatus[] containedFiles = fs.listStatus(fileStatus.getPath());
+            for (FileStatus containedStatus : containedFiles) {
+                // If this dir matches the regex, add all files under it as a split (not include the
```
plase remove these comments since it's out of date

luoyuxia replied:
At least I haven't seen the logic that `If this dir matches the regex,.....`

luoyuxia commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/enumerate/NonSplittingRecursiveAllDirEnumerator.java:
```diff
@@ -0,0 +1,84 @@
+            }
+            final FileStatus[] containedFiles = fs.listStatus(fileStatus.getPath());
+            for (FileStatus containedStatus : containedFiles) {
+                // If this dir matches the regex, add all files under it as a split (not include the
```
dito

luoyuxia commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/FileSystemConnectorOptions.java:
```diff
@@ -68,6 +68,17 @@ public class FileSystemConnectorOptions {
                                     + "The statistics reporting is a heavy operation in some cases,"
                                     + "this config allows users to choose the statistics type according to different situations.");
 
+    public static final ConfigOption<String> SOURCE_REGEX_PATTERN =
```
```suggestion
    public static final ConfigOption<String> SOURCE_PATH_REGEX_PATTERN =
```
?

luoyuxia reviewed (commented):
@ruanhang1993 Thanks for updating. I left minor comments.

luoyuxia reviewed (approved):
@ruanhang1993 Thanks for updating. LGTM.
Btw, please remember to **append** commit next time which make reviewer can track you mofication easier.
May @RamanVerma have a look again?

luoyuxia commented:
Thanks all. Merge it now. Feel free to left comment if you have other concern.
